Improve the low processor mode sleep precision

Calinou opened the pull request:
**Windows 64-bit editor build for testing:** https://0x0.st/i-iB.zip (link expires in late 2020)

The Low Processor Usage Mode Sleep Usec setting is now used as a FPS limiter rather than a constant sleep duration.

This will increase CPU/GPU usage slightly due to the higher effective FPS, but the increase in overall smoothness is worth it. When used in the editor, it will now really cap at ~145 FPS (as it's supposed to) instead of ~130 FPS.

If both Force Fps and Low Processor Usage Mode settings are enabled in the project settings, only the setting that causes the highest sleep duration will be retained.

@Zylann Please test this :slightly_smiling_face:

This closes #11030.

Calinou commented:
@akien-mga Can this be considered for merging for 3.2.2 (as requested [here](https://github.com/godotengine/godot/issues/11030#issuecomment-640964039)), or is it too late now?

akien-mga commented on main/main.cpp:
```diff
@@ -2090,23 +2090,39 @@ bool Main::iteration() {
+	const int target_fps = Engine::get_singleton()->get_target_fps();
+	if ((target_fps > 0 && !Engine::get_singleton()->is_editor_hint()) ||
+			OS::get_singleton()->is_in_low_processor_usage_mode() ||
+			!OS::get_singleton()->can_draw()) {
```
Isn't `!OS::get_singleton()->can_draw()` falling through the cracks of this refactoring with no `time_step`?

akien-mga replied:
I'd suggest refactoring for clarity (first indentation level is function scope):
```cpp
uint64_t dynamic_delay = 0;
if (OS::get_singleton()->is_in_low_processor_usage_mode() || !OS::get_singleton()->can_draw()) {
	dynamic_delay = OS::get_singleton()->get_low_processor_usage_mode_sleep_usec();
}
if (target_fps > 0 && !Engine::get_singleton()->is_editor_hint()) {
	dynamic_delay = MAX(dynamic_delay, (uint64_t)(1000000 / target_fps));
}

if (dynamic_delay > 0) {
	target_ticks += dynamic_delay;
	...
}
```

akien-mga commented:
> Can this be considered for merging in 3.2.2 (as requested here), or is it too late now?

I guess, we can get it tested in 3.2.2 RC 1. Needs a rebase (and possibly further changes, see my review).

Calinou commented:
Now that 3.2.2RC1 has been released, it's probably too late to integrate it :slightly_frowning_face: 

Also, we'd have to move this PR's base branch to `3.2`, as low-processor usage mode isn't currently working in the `master` branch.

akien-mga commented:
> Now that 3.2.2RC1 has been released, it's probably too late to merge it for 3.2.2
> Sorry for the inconvenience.

Well if it's rebased soon I can likely include it in 3.2.2 RC 2, unless you think it's too risky.

> Also, we'd have to move this PR's base branch to `3.2`, as low-processor usage mode isn't currently working in the `master` branch.

It should be merged in `master` nevertheless, even if not working, so that the code is there. If the `3.2` code is too different to easily cherry-pick, you can make a dedicated PR for `3.2`.

Calinou committed:
Improve the low processor mode sleep precision

The Low Processor Usage Mode Sleep Usec setting is now used as a
FPS limiter rather than a constant sleep duration.

This will increase CPU/GPU usage slightly due to the higher
effective FPS, but the increase in overall smoothness is worth it.

If both Force Fps and Low Processor Usage Mode settings are enabled
in the project settings, only the setting that causes the highest
sleep duration will be retained.

This closes #11030.

Calinou commented:
@akien-mga Rebased :slightly_smiling_face: 

I tested the code by backporting it to the `3.2` branch, it works well. You'll need to replace `!DisplayServer::get_singleton()->window_can_draw()` with `!OS::get_singleton()->can_draw()`.

akien-mga commented:
Thanks!
